packages/flutter_tools/lib/src/migrate/migrate_utils.dart

```diff
@@ -0,0 +1,325 @@
+      if (!silent) {
+        logger.printError('Command encountered an error with exit code ${result.exitCode}.');
+        if (commandDescription != null) {
+          logger.printError('Command:');
+          logger.printError(commandDescription, indent: 2);
+        }
+        logger.printError('Stdout:');
+        logger.printStatus(result.stdout as String, indent: 2);
+        logger.printError('Stderr:');
+        logger.printError(result.stderr as String, indent: 2);
+      }
+      if (exit) {
+        throwToolExit('Command failed with exit code ${result.exitCode}', exitCode: result.exitCode);
+      }
+      return false;
+    }
+    return true;
+  }
+
+  /// Returns true if the file does not contain any git conflit markers.
+  static bool conflictsResolved(String contents) {
+    if (contents.contains('>>>>>>>') || contents.contains('=======') || contents.contains('<<<<<<<')) {
+      return false;
+    }
+    return true;
+  }
+}
+
+/// Tracks the output of a git diff command or any special cases such as addition of a new
+/// file or deletion of an existing file.
+class DiffResult {
+  DiffResult(ProcessResult result) :
+    diff = result.stdout as String,
+    isDeletion = false,
+    isAddition = false,
+    isIgnored = false,
+    exitCode = result.exitCode;
+
+  /// Creates a DiffResult that represents a newly added file.
+  DiffResult.addition() :
+    diff = '',
+    isDeletion = false,
+    isAddition = true,
+    isIgnored = false,
+    exitCode = 0;
+
+  /// Creates a DiffResult that represents a deleted file.
+  DiffResult.deletion() :
+    diff = '',
+    isDeletion = true,
+    isAddition = false,
+    isIgnored = false,
+    exitCode = 0;
+
+  /// Creates a DiffResult that represents an ignored file.
+  DiffResult.ignored() :
+    diff = '',
+    isDeletion = false,
+    isAddition = false,
+    isIgnored = true,
+    exitCode = 0;
+
+  /// The diff string output by git.
+  final String diff;
+
+  final bool isDeletion;
+  final bool isAddition;
+  final bool isIgnored;
+
+  /// The exitcode of the command. This is zero when no diffs are found.
+  final int exitCode;
+}
+
+/// Data class to hold the results of a merge.
+class MergeResult {
+  /// Initializes a MergeResult based off of a ProcessResult.
+  MergeResult(ProcessResult result, this.localPath) :
+    mergedString = result.stdout as String,
+    hasConflict = result.exitCode != 0,
+    exitCode = result.exitCode;
+
+  /// Manually initializes a MergeResult with explicit values.
+  MergeResult.explicit({
+    this.mergedString,
+    this.mergedBytes,
+    required this.hasConflict,
+    required this.exitCode,
+    required this.localPath,
+  }) : assert(mergedString == null && mergedBytes != null || mergedString != null && mergedBytes == null);
+
+  /// The final merged string.
+  String? mergedString;
+
+  /// If the file was a binary file, then this field is non-null while mergedString is null.
+  Uint8List? mergedBytes;
```

Comment:
What about making `MergeResult` an abstract class, and implement it with either a `StringMergeResult` or `BinaryMergeResult`